README.md
# 👾  WorldXPlore Alpha  🗺
Randomly generated 2D world, built with [Rust](https://www.rust-lang.org) and [WGPU](https://github.com/gfx-rs/wgpu-rs).

| ![](./.github/readme_img.png) | ![](./.github/readme_inv.png) |
:-------------------------:|:-------------------------:

## 🎮  Gameplay
*As of right now, only keyboard controls are supported. Controller and touch controls may be implemented later.*

WorldXPlore is a 🛠 work in progress 🛠 game about exploring a randomly generated world, gathering resources, and surviving.

WASD controls can be used to move the character around the world. Inventory can be accessed with E, and the game can be paused with ESC, where you can access world saving/loading, and quitting. Interacting with tiles is not yet implemented.

## 🖥  Supported Platforms
| Platform | Status |
|----------|--------|
| Mac OS   | 🤷 Untested, Should Work (^0.2.0), ✅ Tested (0.1.0)|
| Windows  | ✅ Tested|
| Linux    | ✅ Tested (^0.2.0)|
| Android  | ❌ Not Yet Supported |
| iOS      | ❌ Not Yet Supported |

Compiled MacOS bundles in .app format are available under [releases](https://github.com/jamesbmadden/worldxplore/releases). Windows and Linux should be able to compile, but it is not tested.

CHANGELOG.md
## 0.2.0
- It's been a while! There was a lot to catch up on with wgpu since the last time this project was worked on, so this update focuses on bringing wpgu up from version 0.8.0 to 0.17.0 (!)
- This update prepares the project for future updates with new content.
## 0.1.0
- Initial version of the game, written in 2021

src/ui.rs
use crate::render;
use crate::tiles;

use std::convert::TryInto;

pub trait GenVertices {
  fn gen_vertices (&self) -> Vec<render::Vertex>;
}

pub const TILE_WIDTH: f32 = 8. / render::TILESET_WIDTH as f32;
pub const TILE_HEIGHT: f32 = 8. / render::TILESET_HEIGHT as f32;

pub struct Group {
  pub children: Vec<Vec<render::Vertex>> // will later be changed to actually include any object
}

impl Group {

  pub fn gen_vertices (&self) -> Vec<render::Vertex> {
    self.children.iter().flat_map(|child| -> Vec<render::Vertex> {child.iter().cloned().collect()}).collect()
  }
  
}

pub struct Button<F> where F: FnMut() {
  pub label: String,
  pub pos: [f32; 2],
  pub click: F
}

impl<F> Button<F> where F: FnMut() {

  pub fn gen_vertices (&mut self, mouse_pos: &[f32; 2], mouse_down: bool) -> Vec<render::Vertex> {

    let mut vertices: Vec<render::Vertex> = Vec::new();
    // add button texture at specified position
    let start_x = self.pos[0] - TILE_WIDTH * 3.;
    let start_y = self.pos[1] + TILE_HEIGHT / 1.5;
    let end_x = start_x + TILE_WIDTH * 6.;
    let end_y = start_y - TILE_HEIGHT * 1.5;
    let mut tex_coords: [[f32; 2]; 4] = [
      [TILE_WIDTH * 6., TILE_HEIGHT * 4.],
      [TILE_WIDTH * 6., TILE_HEIGHT * 5.],
      [TILE_WIDTH * 10., TILE_HEIGHT * 5.],
      [TILE_WIDTH * 10., TILE_HEIGHT * 4.]
    ];


    // check if mouse is over button
    if start_x <= mouse_pos[0] && mouse_pos[0] <= end_x && start_y >= mouse_pos[1] && mouse_pos[1] >= end_y {
      // move the tex coords over to hover style
      tex_coords[0][0] += TILE_WIDTH * 4.;
      tex_coords[1][0] += TILE_WIDTH * 4.;
      tex_coords[2][0] += TILE_WIDTH * 4.;
      tex_coords[3][0] += TILE_WIDTH * 4.;
      // check if the button is being clicked
      if mouse_down {
        // run the closure
        (self.click)();
      }
    }

    vertices.push(render::Vertex { pos: [ start_x, start_y ], tex_coords: tex_coords[0], animation_frames: 1. }); // top left
    vertices.push(render::Vertex { pos: [ start_x, end_y ], tex_coords: tex_coords[1], animation_frames: 1. }); // bottom left
    vertices.push(render::Vertex { pos: [ end_x, end_y ], tex_coords: tex_coords[2], animation_frames: 1. }); // bottom right
    vertices.push(render::Vertex { pos: [ start_x, start_y ], tex_coords: tex_coords[0], animation_frames: 1. }); // top left
    vertices.push(render::Vertex { pos: [ end_x, end_y ], tex_coords: tex_coords[2], animation_frames: 1. }); // bottom right
    vertices.push(render::Vertex { pos: [ end_x, start_y ], tex_coords: tex_coords[3], animation_frames: 1. }); // top right
    
    // add label at same position
    vertices.append(&mut Label { pos: self.pos, text: self.label.clone(), size_x: TILE_WIDTH / 2., size_y: TILE_HEIGHT / 2. }.gen_vertices());
    // return vertices cloned
    vertices.iter().cloned().collect()
  }

}

pub struct Label {
  pub text: String,
  pub pos: [f32; 2],
  pub size_x: f32,
  pub size_y: f32
}

impl Label {

  pub fn write (text: String, pos: [f32; 2], size_x: f32, size_y: f32) -> (Vec<render::Vertex>, Vec<u16>) {
    let vertices = Label { text, pos, size_x, size_y }.gen_vertices();
    let indices: Vec<u16> = (0_u16..(vertices.len() - 1) as u16).collect();

    (vertices.iter().cloned().collect(), indices.iter().cloned().collect())
  }

  pub fn gen_vertices (&self) -> Vec<render::Vertex> {
    
    let mut vertices: Vec<render::Vertex> = Vec::new();
    let string_length = self.text.len();
    let start_x: f32 = self.pos[0] - (string_length as f32 * self.size_x / 2.);
    let start_y: f32 = self.pos[1] + self.size_y / 2.;
    // iterate through the characters in our string and return text as vertices
    for (i, character) in self.text.chars().enumerate() {
      // get the character position on screen
      let char_x = start_x + self.size_x * i as f32;
      let char_y = start_y;
      // get the texture position in the tileset
      let tex_position = match character {
        'A' => [TILE_WIDTH * 2., TILE_HEIGHT * 5.],
        'B' => [TILE_WIDTH * 3., TILE_HEIGHT * 5.],
        'C' => [TILE_WIDTH * 4., TILE_HEIGHT * 5.],
        'D' => [TILE_WIDTH * 5., TILE_HEIGHT * 5.],
        'E' => [TILE_WIDTH * 6., TILE_HEIGHT * 5.],
        'F' => [TILE_WIDTH * 7., TILE_HEIGHT * 5.],
        'G' => [TILE_WIDTH * 8., TILE_HEIGHT * 5.],
        'H' => [TILE_WIDTH * 9., TILE_HEIGHT * 5.],
        'I' => [TILE_WIDTH * 10., TILE_HEIGHT * 5.],
        'J' => [TILE_WIDTH * 11., TILE_HEIGHT * 5.],
        'K' => [TILE_WIDTH * 12., TILE_HEIGHT * 5.],
        'L' => [TILE_WIDTH * 13., TILE_HEIGHT * 5.],
        'M' => [TILE_WIDTH * 14., TILE_HEIGHT * 5.],
        'N' => [TILE_WIDTH * 15., TILE_HEIGHT * 5.],
        'O' => [TILE_WIDTH * 2., TILE_HEIGHT * 6.],
        'P' => [TILE_WIDTH * 3., TILE_HEIGHT * 6.],
        'Q' => [TILE_WIDTH * 4., TILE_HEIGHT * 6.],
        'R' => [TILE_WIDTH * 5., TILE_HEIGHT * 6.],
        'S' => [TILE_WIDTH * 6., TILE_HEIGHT * 6.],
        'T' => [TILE_WIDTH * 7., TILE_HEIGHT * 6.],
        'U' => [TILE_WIDTH * 8., TILE_HEIGHT * 6.],
        'V' => [TILE_WIDTH * 9., TILE_HEIGHT * 6.],
        'W' => [TILE_WIDTH * 10., TILE_HEIGHT * 6.],
        'X' => [TILE_WIDTH * 11., TILE_HEIGHT * 6.],
        'Y' => [TILE_WIDTH * 12., TILE_HEIGHT * 6.],
        'Z' => [TILE_WIDTH * 13., TILE_HEIGHT * 6.],
        'a' => [0., TILE_HEIGHT * 7.],
        'b' => [TILE_WIDTH, TILE_HEIGHT * 7.],
        'c' => [TILE_WIDTH * 2., TILE_HEIGHT * 7.],
        'd' => [TILE_WIDTH * 3., TILE_HEIGHT * 7.],
        'e' => [TILE_WIDTH * 4., TILE_HEIGHT * 7.],
        'f' => [TILE_WIDTH * 5., TILE_HEIGHT * 7.],
        'g' => [TILE_WIDTH * 6., TILE_HEIGHT * 7.],
        'h' => [TILE_WIDTH * 7., TILE_HEIGHT * 7.],
        'i' => [TILE_WIDTH * 8., TILE_HEIGHT * 7.],
        'j' => [TILE_WIDTH * 9., TILE_HEIGHT * 7.],
        'k' => [TILE_WIDTH * 10., TILE_HEIGHT * 7.],
        'l' => [TILE_WIDTH * 11., TILE_HEIGHT * 7.],
        'm' => [TILE_WIDTH * 12., TILE_HEIGHT * 7.],
        'n' => [TILE_WIDTH * 13., TILE_HEIGHT * 7.],
        'o' => [TILE_WIDTH * 14., TILE_HEIGHT * 7.],
        'p' => [TILE_WIDTH * 15., TILE_HEIGHT * 7.],
        'q' => [0., TILE_HEIGHT * 8.],
        'r' => [TILE_WIDTH, TILE_HEIGHT * 8.],
        's' => [TILE_WIDTH * 2., TILE_HEIGHT * 8.],
        't' => [TILE_WIDTH * 3., TILE_HEIGHT * 8.],
        'u' => [TILE_WIDTH * 4., TILE_HEIGHT * 8.],
        'v' => [TILE_WIDTH * 5., TILE_HEIGHT * 8.],
        'w' => [TILE_WIDTH * 6., TILE_HEIGHT * 8.],
        'x' => [TILE_WIDTH * 7., TILE_HEIGHT * 8.],
        'y' => [TILE_WIDTH * 8., TILE_HEIGHT * 8.],
        'z' => [TILE_WIDTH * 9., TILE_HEIGHT * 8.],
        '0' => [TILE_WIDTH * 10., TILE_HEIGHT * 8.],
        '1' => [TILE_WIDTH * 11., TILE_HEIGHT * 8.],
        '2' => [TILE_WIDTH * 12., TILE_HEIGHT * 8.],
        '3' => [TILE_WIDTH * 13., TILE_HEIGHT * 8.],
        '4' => [TILE_WIDTH * 14., TILE_HEIGHT * 8.],
        '5' => [TILE_WIDTH * 15., TILE_HEIGHT * 8.],
        '6' => [0., TILE_HEIGHT * 9.],
        '7' => [TILE_WIDTH, TILE_HEIGHT * 9.],
        '8' => [TILE_WIDTH * 2., TILE_HEIGHT * 9.],
        '9' => [TILE_WIDTH * 3., TILE_HEIGHT * 9.],
        _ => [TILE_WIDTH, TILE_HEIGHT * 4.]
      };
      // now we can generate the vertices  - we don't give indices because it is up to the root UI element to just
      // make a list based on a (0..(vertices.len() - 1)).collect() so that multiple UI Elements can be easily combined
      // first tri: top left bottom left bottom right
      vertices.push(render::Vertex { pos: [char_x, char_y], tex_coords: tex_position, animation_frames: 1. }); // top left
      vertices.push(render::Vertex { pos: [char_x, char_y - self.size_y], tex_coords: [tex_position[0], tex_position[1] + TILE_HEIGHT], animation_frames: 1. }); // bottom left
      vertices.push(render::Vertex { pos: [char_x + self.size_x, char_y - self.size_y], tex_coords: [tex_position[0] + TILE_WIDTH, tex_position[1] + TILE_HEIGHT], animation_frames: 1. }); // bottom right
      // second tri: top left bottom right top right
      vertices.push(render::Vertex { pos: [char_x, char_y], tex_coords: tex_position, animation_frames: 1. }); // top left
      vertices.push(render::Vertex { pos: [char_x + self.size_x, char_y - self.size_y], tex_coords: [tex_position[0] + TILE_WIDTH, tex_position[1] + TILE_HEIGHT], animation_frames: 1. }); // bottom right
      vertices.push(render::Vertex { pos: [char_x + self.size_x, char_y], tex_coords: [tex_position[0] + TILE_WIDTH, tex_position[1]], animation_frames: 1. }); // top right

    }

    vertices.iter().cloned().collect()

  }

}

pub struct Inventory<'a> {
  pub pos: [f32; 2],
  pub size_x: f32,
  pub size_y: f32,
  pub inventory: &'a Vec<tiles::ItemProperties<'a>>
}

impl Inventory<'_> {

  pub fn gen_vertices (&self, mouse_pos: &[f32; 2]) -> Vec<render::Vertex> {
    let mut vectors: Vec<render::Vertex> = Vec::new();
    let items_per_row: usize = 8;
    // start position
    let start_x: f32 = 0.8 / (items_per_row as f32 * 3.) - 0.75;
    let start_y: f32 = self.pos[1] as f32;

    for (i, item) in self.inventory.iter().enumerate() {
      // find the column that it is in
      let column = i % items_per_row;
      let row = i / items_per_row;
      // make the background
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.), start_y - (row as f32 * TILE_HEIGHT * 3. ) ], tex_coords: [TILE_WIDTH * 5., TILE_HEIGHT * 4.], animation_frames: 1. }); // top left
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.), start_y - (row as f32 * TILE_HEIGHT * 3. ) + TILE_HEIGHT * 1.5 ], tex_coords: [TILE_WIDTH * 5., TILE_HEIGHT * 5.], animation_frames: 1. }); // bottom left
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.) + TILE_WIDTH * 1.5, start_y - (row as f32 * TILE_HEIGHT * 3. ) + TILE_HEIGHT * 1.5 ], tex_coords: [TILE_WIDTH * 6., TILE_HEIGHT * 5.], animation_frames: 1. }); // bottom right
      // second tri
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.), start_y - (row as f32 * TILE_HEIGHT * 3. ) ], tex_coords: [TILE_WIDTH * 5., TILE_HEIGHT * 4.], animation_frames: 1. }); // top left
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.) + TILE_WIDTH * 1.5, start_y - (row as f32 * TILE_HEIGHT * 3. ) + TILE_HEIGHT * 1.5 ], tex_coords: [TILE_WIDTH * 6., TILE_HEIGHT * 5.], animation_frames: 1. }); // bottom right
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.) + TILE_WIDTH * 1.5, start_y - (row as f32 * TILE_HEIGHT * 3. ) ], tex_coords: [TILE_WIDTH * 5., TILE_HEIGHT * 5.], animation_frames: 1. }); // top right

      // make the icon
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.), start_y - (row as f32 * TILE_HEIGHT * 3. ) ], tex_coords: [TILE_WIDTH * item.ts_coord_x as f32, TILE_HEIGHT * (item.ts_coord_y as f32 + 1.)], animation_frames: 1. }); // top left
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.), start_y - (row as f32 * TILE_HEIGHT * 3. ) + TILE_HEIGHT * 1.5 ], tex_coords: [TILE_WIDTH * item.ts_coord_x as f32, TILE_HEIGHT * (item.ts_coord_y as f32)], animation_frames: 1. }); // bottom left
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.) + TILE_WIDTH * 1.5, start_y - (row as f32 * TILE_HEIGHT * 3. ) + TILE_HEIGHT * 1.5 ], tex_coords: [TILE_WIDTH * (item.ts_coord_x as f32 + 1.), TILE_HEIGHT * (item.ts_coord_y as f32)], animation_frames: 1. }); // bottom right
      // second tri
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.), start_y - (row as f32 * TILE_HEIGHT * 3. ) ], tex_coords: [TILE_WIDTH * item.ts_coord_x as f32, TILE_HEIGHT * (item.ts_coord_y as f32 + 1.)], animation_frames: 1. }); // top left
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.) + TILE_WIDTH * 1.5, start_y - (row as f32 * TILE_HEIGHT * 3. ) + TILE_HEIGHT * 1.5 ], tex_coords: [TILE_WIDTH * (item.ts_coord_x as f32 + 1.), TILE_HEIGHT * (item.ts_coord_y as f32)], animation_frames: 1. }); // bottom right
      vectors.push(render::Vertex { pos: [ start_x + (column as f32 * TILE_WIDTH * 3.) + TILE_WIDTH * 1.5, start_y - (row as f32 * TILE_HEIGHT * 3. ) ], tex_coords: [TILE_WIDTH * (item.ts_coord_x as f32 + 1.), TILE_HEIGHT * (item.ts_coord_y as f32 + 1.)], animation_frames: 1. }); // top right

      // check if label should be shown (on mouse hover)
      //if (mouse_pos[0] >= start_x + (column as f32 * TILE_WIDTH * 3.)) && (start_x + (column as f32 * TILE_WIDTH * 3.) + TILE_WIDTH >= mouse_pos[0]) && (mouse_pos[1] <= start_y - (row as f32 * TILE_HEIGHT * 3. )) && (start_y - (row as f32 * TILE_HEIGHT * 3. ) + TILE_HEIGHT <= mouse_pos[1]) {
        vectors.append(&mut Label { text: String::from(item.name), size_x: self.size_x / 6., size_y: self.size_y / 6., pos: [start_x + (column as f32 * TILE_WIDTH * 3.) + TILE_WIDTH * 0.5, start_y - (row as f32 * TILE_HEIGHT * 3. ) - TILE_HEIGHT * 0.5 ]}.gen_vertices());
      //}
    }

    vectors.iter().cloned().collect()
  }

}

src/worldgen.rs
use noise::{NoiseFn, Perlin, Seedable};
use crate::tiles;

/**
* Create a map of specified size with perlin noise points
*/
pub fn generate_perlin (width: i32, height: i32, seed: u32) -> Vec<Vec<(f64, f64)>> {
  // create a vector to store world data in
  let mut world: Vec<Vec<(f64, f64)>> = Vec::new();
  // make a perlin noise function to read from
  let p_noise = Perlin::new().set_seed(seed);
  // noise for vegetation
  let veg_noise = Perlin::new().set_seed(p_noise.seed() + 1);
  println!("Seed: {}", p_noise.seed());
  // loop [x][y] the size specified
  for x in 0..width {
    let mut row: Vec<(f64, f64)> = Vec::new();
    for y in 0..height {
      row.push((p_noise.get([x as f64 / 16., y as f64 / 16., 0.]), veg_noise.get([x as f64 / 1.6, y as f64 / 1.6, 0.])));
    }
    world.push(row.iter().cloned().collect());
  }

  world.iter().cloned().collect()
}

/**
* Convert perlin noise map to game tiles
* 0: Deep Ocean
* 1: Ocean
* 2: Beach
* 3: Grass
* 4: Trees
*/
pub fn elevation_to_tiles (p_map: Vec<Vec<(f64, f64)>>) -> Vec<Vec<tiles::TileProperties>> {
  // map over the map
  p_map.iter().map(|p_row| p_row.iter().map(|(tile, veg)| {
    if tile < &-0.3 { 
      if veg > &0.5 { return tiles::KELP }
      else { return tiles::DEEP_OCEAN }
    }
    else if tile < &0. { 
      if veg > &0.7 { return tiles::LILYPAD }
      else { return tiles::OCEAN }
    }
    else if tile < &0.2 { 
      if veg > &0.7 { return tiles::CACTUS }
      else { return tiles::SAND }
    }
    else { 
      if veg > &0.7 { return tiles::STUMP }
      else if veg > &0.5 { return tiles::TREE }
      else { return tiles::GRASS }
    }
  }).collect()).collect()
}

src/player.rs
use std::collections::HashSet;
use std::convert::TryInto;
use std::fs;
use bytemuck::{Pod, Zeroable};
use serde::{Serialize, Deserialize};
use crate::tiles;
use crate::render; // for the tileset size constants
use crate::ui;
use crate::worldgen;

use std::f32::consts::PI;

pub struct Player<'a> {
  pub keys_down: HashSet<winit::event::VirtualKeyCode>,
  pub width: i32,
  pub height: i32,
  pub x: f32,
  pub y: f32,
  pub x_speed: f32,
  pub y_speed: f32,
  pub uniforms: Uniforms,
  pub is_swimming: bool,
  pub health: f32,
  pub max_health: f32,
  pub paused: bool,
  pub pause_type: PauseType,
  pub seed: u32,
  pub world_name: String,
  pub inventory: Vec<tiles::ItemProperties<'a>>
}

#[derive(Serialize, Deserialize, Debug)]
pub struct GameData {
  pub x: f32,
  pub y: f32,
  pub health: f32,
  pub max_health: f32,
  pub seed: u32,
  pub time: f32
}

/**
 * The various ways the game can be paused.
 */
pub enum PauseType {
  Pause,
  Inventory
}

impl Player<'_> {

  pub fn new (width: i32, height: i32, seed: u32) -> Self {
    Player {
      keys_down: HashSet::new(),
      x: 0., y: 0., x_speed: 0., y_speed: 0.,
      width, height, seed,
      uniforms: Uniforms::default(),
      is_swimming: false,
      health: 6., max_health: 6.,
      paused: false,
      pause_type: PauseType::Pause,
      world_name: String::from("New Game"),
      inventory: vec![tiles::SWORD, tiles::STICK]
    }
  }

  // if keys are pressed, update x and y values
  pub fn update (&mut self, world: &mut Vec<Vec<tiles::TileProperties>>, cam_width: i32, cam_height: i32) {

    self.width = cam_width;
    self.height = cam_height;
    
    // only update the player position if the game is paused
    if !self.paused {
      self.move_character(world);
      self.uniforms.time += 0.01;
    }

    // make transforms for how much to offset tiles for smoother scrolling
    let x_offset = self.x as f32 % 1.;
    let y_offset = self.y as f32 % 1.;
    // coordinate space is -1.0 to 1.0, so size should be double as big as 1 / size
    let tile_width = 2. / self.width as f32;
    let tile_height = 2. / self.height as f32;
    // set as uniforms to be rendered with
    self.uniforms = Uniforms { 
      translate_vector: [ x_offset * tile_width, y_offset * tile_height ], 
      is_swimming: self.is_swimming.into(),
      time: self.uniforms.time,
      light_intensity: light_intensity(self.uniforms.time),
      unused: 0.
    };

  }

  pub fn move_character (&mut self, world: &Vec<Vec<tiles::TileProperties>>) {
    // movement speed stuff
    if self.is_swimming {
      self.x_speed /= 1.8;
      self.y_speed /= 1.8;
    } else {
      self.x_speed /= 1.2;
      self.y_speed /= 1.2;
    }

    if self.keys_down.contains(&winit::event::VirtualKeyCode::A) {
      self.x_speed -= 0.02;
    }
    if self.keys_down.contains(&winit::event::VirtualKeyCode::D) {
      self.x_speed += 0.02;
    }
    if self.keys_down.contains(&winit::event::VirtualKeyCode::S) {
      self.y_speed += 0.02;
    }
    if self.keys_down.contains(&winit::event::VirtualKeyCode::W) {
      self.y_speed -= 0.02;
    }
    // attempt to move
    self.x += self.x_speed;
    // check if you can actually move to the position
    if world[(self.x + 0.5).floor() as usize + (self.width / 2) as usize][self.y.floor() as usize + (self.height / 2) as usize + 2].solid {
      self.x -= self.x_speed;
    }
    // same thing as above
    self.y += self.y_speed;
    if world[(self.x + 0.5).floor() as usize + (self.width / 2) as usize][self.y.floor() as usize + (self.height / 2) as usize + 2].solid {
      self.y -= self.y_speed;
    }

    // set in water to false before checking if we are in water
    self.is_swimming = false;
    // if move lands us in water, set in water to true
    // adjustments are made to set the comparison point at the centre of the character's feet instead of
    // the top left of the screen
    // values <= 1 are water varities
    if world[(self.x + 0.5).floor() as usize + (self.width / 2) as usize][self.y.floor() as usize + (self.height / 2) as usize + 2].swimmable {
      self.is_swimming = true;
    }

    // prevent from raising or lowering the x or y past bounds
    if self.x < 0. {
      self.x = 0.;
    }
    if self.y < 0. {
      self.y = 0.;
    }
  }

  // key pressed, add it to keys down
  pub fn key_pressed (&mut self, key: winit::event::VirtualKeyCode) {
    self.keys_down.insert(key);
  }
  // key released, add it to keys down
  pub fn key_released (&mut self, key: winit::event::VirtualKeyCode) {
    self.keys_down.remove(&key);
  }

  pub fn gen_ui_vertices (&mut self, mouse_pos: [f32; 2], mouse_down: bool, control_flow: &mut winit::event_loop::ControlFlow, mut world: &mut Vec<Vec<tiles::TileProperties>>) -> (Vec<render::Vertex>, Vec<u16>) {

    let mut vertices: Vec<render::Vertex> = Vec::new();
    let mut indices: Vec<u16> = Vec::new();
    // create variables for texture locations
    let tile_width: f32 = 4. / self.width as f32;
    let tile_height: f32 = 4. / self.height as f32;
    let texture_width: f32 = 8. / render::TILESET_WIDTH as f32;
    let texture_height: f32 = 8. / render::TILESET_HEIGHT as f32;

    // generate health vertices
    for heart in 0..((self.max_health / 2.).ceil() as i32) {
      // get positions for heart
      let heart_x: f32 = -1.0 + tile_width * heart as f32 + (tile_width / 10.);
      let heart_y: f32 = 1.0 - tile_height / 10.;
      // add the vertices
      vertices.push(render::Vertex { pos: [ heart_x, heart_y ], tex_coords: [ texture_width * 2., texture_height * 4. ], animation_frames: 1.}); // top left
      vertices.push(render::Vertex { pos: [ heart_x, heart_y - tile_height * 9. / 10. ], tex_coords: [ texture_width * 2., texture_height * 5. ], animation_frames: 1.}); // bottom left
      vertices.push(render::Vertex { pos: [ heart_x + tile_width * 9. / 10., heart_y - tile_height * 9. / 10. ], tex_coords: [ texture_width * 3., texture_height * 5. ], animation_frames: 1.}); // bottom right
      vertices.push(render::Vertex { pos: [ heart_x + tile_width * 9. / 10., heart_y ], tex_coords: [ texture_width * 3., texture_height * 4. ], animation_frames: 1.}); // top right
      // add in the indices
      let len = vertices.len();
      indices.append(&mut vec![ (len - 4).try_into().unwrap(), (len - 3).try_into().unwrap(), (len - 2).try_into().unwrap(), (len - 4).try_into().unwrap(), (len - 2).try_into().unwrap(), (len - 1).try_into().unwrap() ]);
    }

    // if game is paused, add text and buttons
    if self.paused {

      // black background
      vertices.push(render::Vertex { pos: [ -1., 1. ], tex_coords: [ texture_width * 5., texture_height * 4. ], animation_frames: 1.}); // top left
      vertices.push(render::Vertex { pos: [ -1., -1. ], tex_coords: [ texture_width * 5., texture_height * 5. ], animation_frames: 1.}); // bottom left
      vertices.push(render::Vertex { pos: [ 1., -1. ], tex_coords: [ texture_width * 6., texture_height * 5. ], animation_frames: 1.}); // bottom right
      vertices.push(render::Vertex { pos: [ 1., 1. ], tex_coords: [ texture_width * 6., texture_height * 4. ], animation_frames: 1.}); // top right
      // add in the indices
      let len = vertices.len();
      indices.append(&mut vec![ (len - 4).try_into().unwrap(), (len - 3).try_into().unwrap(), (len - 2).try_into().unwrap(), (len - 4).try_into().unwrap(), (len - 2).try_into().unwrap(), (len - 1).try_into().unwrap() ]);

      let mut pause_ui_vertices: Vec<render::Vertex> = Vec::new();

      pause_ui_vertices = match self.pause_type {
        PauseType::Pause => {

          // The pause title
          ui::Group {
            children: vec![
              ui::Label { pos: [0., 0.5], text: String::from("Paused"), size_x: tile_width, size_y: tile_height }.gen_vertices(),
              ui::Button { pos: [0., 0.], label: String::from("Resume"), click: || { self.paused = false; } }.gen_vertices(&mouse_pos, mouse_down),
              ui::Button { pos: [0., -0.2], label: String::from("Save Game"), click: || { self.write_out_gamedata(); } }.gen_vertices(&mouse_pos, mouse_down),
              ui::Button { pos: [0., -0.4], label: String::from("Load Game"), click: || { self.load_gamedata( self.read_gamedata()); self.paused = false; } }.gen_vertices(&mouse_pos, mouse_down),
              ui::Button { pos: [0., -0.6], label: String::from("Quit"), click: || { *control_flow = winit::event_loop::ControlFlow::Exit; } }.gen_vertices(&mouse_pos, mouse_down)
            ]
          }.gen_vertices()
          
        },
        PauseType::Inventory => {
          ui::Group {
            children: vec![
              ui::Label { pos: [0., 0.75], text: String::from("Inventory"), size_x: tile_width, size_y: tile_height }.gen_vertices(),
              ui::Inventory { pos: [0., 0.25], size_x: tile_width, size_y: tile_height, inventory: &self.inventory }.gen_vertices(&mouse_pos)
            ]
          }.gen_vertices()
        },
        _ => {
          ui::Label { pos: [0., 0.], text: String::from("Error"), size_x: tile_width, size_y: tile_height }.gen_vertices()
        }
      };

      let index_start: u16 = vertices.len().try_into().unwrap();
      let pause_label_length: u16 = pause_ui_vertices.len().try_into().unwrap();
      let index_end: u16 = index_start + pause_label_length;
      vertices.append(&mut pause_ui_vertices);

      for n in index_start..index_end {
        indices.push(n);
      }
    }

    ( vertices.iter().cloned().collect(), indices.iter().cloned().collect() )
  }

  pub fn write_out_gamedata (&self) {
    let gamedata = GameData { health: self.health, max_health: self.max_health, seed: self.seed, x: self.x, y: self.y, time: self.uniforms.time };
    let file_string = serde_yaml::to_string(&gamedata).unwrap();
    fs::create_dir("worlds").unwrap_or(());
    fs::write(format!("worlds/{}.yaml", self.world_name), &file_string).unwrap();
  }
  pub fn read_gamedata (&self) -> GameData {
    let file_string = fs::read_to_string(format!("worlds/{}.yaml", self.world_name)).unwrap();
    fs::create_dir("worlds").unwrap_or(());
    let result: GameData = serde_yaml::from_str(&file_string).unwrap();
    println!("{:?}", result);
    result
  }
  pub fn load_gamedata (&mut self, gamedata: GameData) -> Vec<Vec<tiles::TileProperties>> {
    self.x = gamedata.x;
    self.y = gamedata.y;
    self.max_health = gamedata.max_health;
    self.health = gamedata.health;
    self.uniforms.time = gamedata.time;
    worldgen::elevation_to_tiles(worldgen::generate_perlin(1000, 1000, gamedata.seed))
  }

}


#[repr(C)]
#[derive(Clone, Copy, Pod, Zeroable, Debug)]
pub struct Uniforms {
  pub translate_vector: [f32; 2],
  pub is_swimming: i32,
  pub time: f32,
  pub light_intensity: [f32; 3],
  // needed to fix an issue with the shader that I'm not sure about
  pub unused: f32
}

impl Uniforms {
  pub fn default() -> Self {
    Uniforms {
      translate_vector: [0., 0.],
      is_swimming: 0,
      time: 0.,
      light_intensity: light_intensity(0.),
      unused: 0.
    }
  }
}

// create a light intensity for the shader based on current time
pub fn light_intensity (time: f32) -> [f32; 3] {

  let day_length: f32 = 36.; // length of the day (currently a minute)

  let offset_time = 2. * PI / day_length * time; // so that the cycle happens every length of day instead of pi

  let r: f32 = 1_f32.min(0.03_f32.max(offset_time.sin() + 1.)); // r value; sliced sine
  let g: f32 = 1_f32.min(0.06_f32.max(offset_time.sin() + 0.9)); // g value; sliced sine, make it rise slower for sunrise / sunset
  let b: f32 = 1_f32.min(0.1_f32.max(offset_time.sin() + 0.8)); // b value; sliced sine, make it rise slowest for sunrise / sunset

  return [r, g, b];
}

// return the vertices and indices to form player sprite
pub fn player_vertices (width: i32, height: i32) -> (Vec<render::Vertex>, Vec<u16>) {

  let tile_width: f32 = 1. / width as f32;
  let tile_height: f32 = 1. / height as f32;
  let texture_width: f32 = 8. / render::TILESET_WIDTH as f32;
  let texture_height: f32 = 8. / render::TILESET_HEIGHT as f32;
  // player data:
  ( 
    vec![ // player vertices - texture coords depend on whether in water or not for different texture
      render::Vertex { pos: [ -tile_width, tile_height * 3. ], tex_coords: [ 0., texture_height * 4. ], animation_frames: 1. }, // top left
      render::Vertex { pos: [ -tile_width, tile_height * -3. ], tex_coords: [ 0., texture_height * 7. ], animation_frames: 1. }, // bottom left
      render::Vertex { pos: [ tile_width, tile_height * -3. ], tex_coords: [ texture_width, texture_height * 7. ], animation_frames: 1. }, // bottom right
      render::Vertex { pos: [ tile_width, tile_height * 3. ], tex_coords: [ texture_width, texture_height * 4. ], animation_frames: 1. } // top right
    ],
    vec![ 0, 1, 2, 0, 2, 3 ] // player indices
  )
}